File: src/main/java/backend/academy/flame/task/ApplyGammaTask.java
package backend.academy.flame.task;

import backend.academy.flame.GammaCorrector;
import backend.academy.pixel.Pixel;
import java.util.concurrent.RecursiveAction;
import lombok.RequiredArgsConstructor;

/**
 * Задача применения гамма-коррекция, способная делиться на части
 */
@RequiredArgsConstructor
public class ApplyGammaTask extends RecursiveAction implements GammaCorrector {
    private static final int THRESHOLD = 50; // Минимальное количество строк для разделения задачи
    private final Pixel[][] pixels;
    private final double[][] normals;
    private final double gamma;
    private final double maxNormal;
    private final int startRow;
    private final int endRow;

    public ApplyGammaTask(Pixel[][] pixels, double[][] normals, double gamma, double maxNormal) {
        this(pixels, normals, gamma, maxNormal, 0, pixels.length);
    }

    @Override
    protected void compute() {
        if (endRow - startRow <= THRESHOLD) {
            // Обрабатываем строки
            for (int i = startRow; i < endRow; i++) {
                for (int j = 0; j < pixels[i].length; j++) {
                    correct(pixels, i, j, gamma, maxNormal, normals);
                }
            }
        } else {
            // Разделяем задачу
            int mid = (startRow + endRow) / 2;
            ApplyGammaTask topTask = new ApplyGammaTask(pixels, normals, gamma, maxNormal, startRow, mid);
            ApplyGammaTask bottomTask = new ApplyGammaTask(pixels, normals, gamma, maxNormal, mid, endRow);
            invokeAll(topTask, bottomTask);
        }
    }
}

File: src/main/kotlin/backend/academy/flame/task/ApplyGammaTask.kt
import backend.academy.flame.GammaCorrector
import backend.academy.pixel.Pixel
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch

/**
 * Задача применения гамма-коррекции, способная делиться на части
 */
class ApplyGammaTask(
    private val pixels: Array<Array<Pixel?>>,
    private val normals: Array<DoubleArray>,
    private val gamma: Double,
    private val maxNormal: Double,
    private val startRow: Int = 0,
    private val endRow: Int = pixels.size
) : GammaCorrector {

     suspend fun compute() {
        if (endRow - startRow <= THRESHOLD) {
            // Обрабатываем строки
            for (i in startRow until endRow) {
                for (j in pixels[i].indices) {
                    correct(pixels, i, j, gamma, maxNormal, normals)
                }
            }
        } else {
            // Разделяем задачу
            val mid = (startRow + endRow) / 2
            val topTask = ApplyGammaTask(pixels, normals, gamma, maxNormal, startRow, mid)
            val bottomTask = ApplyGammaTask(pixels, normals, gamma, maxNormal, mid, endRow)
            coroutineScope {
                val job1 = launch { topTask.compute() }
                val job2 = launch { bottomTask.compute() }
                job1.join()
                job2.join()
            }
        }
    }


    companion object {
        private const val THRESHOLD = 50 // Минимальное количество строк для разделения задачи
    }
}